Framework/Algorithms/src/CalculatePlaczekSelfScattering.cpp

```diff
@@ -0,0 +1,163 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "MantidAlgorithms/CalculatePlaczekSelfScattering.h"
+#include "MantidAPI/Sample.h"
+#include "MantidAPI/WorkspaceFactory.h"
+#include "MantidDataObjects/Workspace2D.h"
+#include "MantidDataObjects/WorkspaceCreation.h"
+#include "MantidGeometry/Instrument/DetectorInfo.h"
+#include "MantidKernel/Atom.h"
+#include "MantidKernel/Material.h"
+
+#include <utility>
+
+namespace Mantid {
+namespace Algorithms {
+
+DECLARE_ALGORITHM(CalculatePlaczekSelfScattering)
+
+void CalculatePlaczekSelfScattering::init() {
+  declareProperty(
+      std::make_unique<API::WorkspaceProperty<API::MatrixWorkspace>>(
+          "InputWorkspace", "", Kernel::Direction::Input),
+      "Workspace of fitted incident spectrum with it's first derivative. "
+      "Workspace must have instument and sample data.");
+  declareProperty(
+      std::make_unique<API::WorkspaceProperty<API::MatrixWorkspace>>(
+          "OutputWorkspace", "", Kernel::Direction::Output),
+      "Workspace with the Self scattering correction");
+}
+//----------------------------------------------------------------------------------------------
+/** Validate inputs.
+ */
+std::map<std::string, std::string>
+CalculatePlaczekSelfScattering::validateInputs() {
+  std::map<std::string, std::string> issues;
+  const API::MatrixWorkspace_sptr inWS = getProperty("InputWorkspace");
+  const Geometry::DetectorInfo det_info = inWS->detectorInfo();
+  if (det_info.size() == 0) {
+    issues["InputWorkspace"] =
+        "Input workspace does not have detector information";
+  }
+  Kernel::Material::ChemicalFormula formula =
+      inWS->sample().getMaterial().chemicalFormula();
+  if (formula.size() == 0) {
+    issues["InputWorkspace"] = "Input workspace does not have a valid sample";
+  }
+  return issues;
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void CalculatePlaczekSelfScattering::exec() {
+  const API::MatrixWorkspace_sptr inWS = getProperty("InputWorkspace");
+  API::MatrixWorkspace_sptr outWS = getProperty("OutputWorkspace");
+  const double factor =
+      1.0 / 1.66053906660e-27; // atomic mass unit-kilogram relationship
+  const double neutron_mass = factor * 1.674927471e-27; // neutron mass
+  // get sample information : mass, total scattering length, and concentration
+  // of each species
+  auto atom_species =
+      CalculatePlaczekSelfScattering::get_sample_species_info(inWS);
+  // calculate summation term w/ neutron mass over molecular mass ratio
+  double summation_term = 0.0;
```

Comment:
This could be done with `std::accumulate`, I'll leave it up to you if you want to change it though.

Reply:
I think it's not worth bothering with that in this case